docs/contracts-intro.md

```diff
@@ -2,11 +2,181 @@
+associated tables is the same for both Common and Core Contracts.
+
+### Common Contracts
+
+All OpenSAFELY backends satisfy Common Contracts.
+
+There will be several different Common Contracts available with the
+initial release of Data Builder. More Common Contracts may be added in
+future.
+
+!!! warning
+    Some backends might deviate slightly from the suggested OpenSAFELY
+    behaviour shared by most backends. Refer to the [OpenSAFELY
+    Contracts Reference](contracts-reference.md) for backends you intend
+    to access with your dataset definition.
+
+For dataset definitions that only access tables defined by Common
+Contracts, those dataset definitions should be largely compatible with all
+OpenSAFELY backends.
+
+### Custom Contracts
+
+Some backends may offer particularly specialised data that is not yet
+covered by a Common Contract. *Custom Contracts* allow individual
+backends to still offer this data via OpenSAFELY.
+
+!!! warning
+    Using Custom Contracts might tie a dataset definition more closely
+    to a specific backend, than if just Common Contracts are used. This
+    also depends on how the dataset definition is designed.
+
+!!! note
+    Should additional data providers want to offer data that closely resembles
+    an existing *Custom* Contract, OpenSAFELY developers would consider a new
+    *Common* Contract.
+
+## Reading an OpenSAFELY Contract
+
+!!! note
+    Refer to the [OpenSAFELY Contracts
+    Reference](contracts-intro#the-opensafely-contracts-reference) for
+    the details of specific contracts.
+
+Each OpenSAFELY Contract relates to a specific domain of health record
+data. The name of the Contract indicates the data domain covered by the
+Contract. Tables that satisfy these Contracts can be accessed in a
+dataset definition via [Data Builder's query language,
+ehrQL](ehrql-intro.md).
+
+An OpenSAFELY Contract provides the following information:
+
+* the **variable names**. Each variable is denoted by the
+  identifier you use to access the variable in ehrQL.
+* the **variable's description**. The description summarises what the
+  variable represents.
+* the **variable's data type**. For instance, a variable
+  might be one of the following:
+    * integer type
+    * floating point type
+    * date type
+    * string type
+    * or some other type not in this list
+* the **additional constraints on the variable's data values**, other
+  than those given by the variable's type. For instance,
+    * the variable might be of date type, but all values are the first
+      of the month
+    * the variable might be of integer type, but all values are positive
+      integers
+
+## Miscellaneous TODOs
+
+* TODO: Should we refer to columns or variables?
+* TODO: Finalise details above based on what we actually display. Do we
+  know all the variable types, or can we link to them?
+* TODO: Do we want to have nulls as a constraint or as a separate
+  column in the reference?
```

Comment:
This is really out of scope for this page, but I think we should have columns for very common things like this (which I would call "required" rather than referring to nulls);